Add documentation about java agent structure and related classloaders (#380)
open-telemetry/opentelemetry-auto-instr-java: CONTRIBUTING.md

These classes should be available from all possible classloaders in the running application. 
For this reason `java-agent` puts all these classes into JVM's bootstrap classloader.
For the same reason this module should be as small as possible and have as few dependencies as possible. 
Otherwise, there is a risk of accidentally exposing this classes to the actual application.

#### `agent-tooling` module and `instrumentation` submodules
Contains everything necessary to make instrumentation machinery work, 
including integration with [ByteBuddy](https://bytebuddy.net/) and actual library-specific instrumentations. 
As these classes depend on many classes from different libraries, 
it is paramount to hide all these classes from the host application. 
This is achieved in the following way:
* When `java-agent` module builds the final agent, it moves all classes from `instrumentation` submodules 
and `agent-tooling` module into a separate folder inside final jar file, called `auto-tooling-and-instrumentation.isolated`. 
In addition, the extension of all class files is changed from `class` to `classdata`. 
This ensures that general classloaders cannot find nor load these classes.

> * When `io.opentelemetry.auto.bootstrap.Agent` starts instrumentation agent, 

Reviewer: not totally sure about this suggestion
```suggestion
* When `io.opentelemetry.auto.bootstrap.Agent` starts up, 
```

it creates an instance of `io.opentelemetry.auto.bootstrap.AgentClassLoader`, 
loads an `io.opentelemetry.auto.tooling.AgentInstaller` from that `AgentClassLoader` 

> and lets it install all instrumentations with the help of ByteBuddy.

Reviewer: ```suggestion
and then passes control on to the `AgentInstaller` (now in the `AgentClassLoader`). The `AgentInstaller` then installs all of the instrumentations with the help of ByteBuddy.
```

The complicated process above ensures that the majority of auto-instrumentation agent's classes
are totally isolated from application classes, 
and an instrumented class from arbitrary classloader in JVM can still access helper classes from bootstrap classloader.

#### Agent jar structure
If you now look inside `java-agent/build/libs/opentelemetry-auto-<version>.jar`, 
you will see the following "clusters" of classes:
* `auto-tooling-and-instrumentation.isolated/` - contains `agent-tooling` module and 
`instrumentation` submodules, loaded and isolated inside `AgentClassLoader`. 

> Including Open Telemetry SDK.

Reviewer: ```suggestion
Including OpenTelemetry SDK.
```

* `io/opentelemetry/auto/bootstrap/` - contains `agent-bootstrap` module and available in
bootstrap classloader.

> * `io/opentelemetry/auto/shaded/` - contains Open Telemetry API and its dependencies. 

Reviewer: ```suggestion
* `io/opentelemetry/auto/shaded/` - contains OpenTelemetry API and its dependencies. 
```
